packages/vscode-extension/src/webview/components/DeviceRenameDialog.tsx

```diff
@@ -0,0 +1,85 @@
+import "./DeviceRenameDialog.css";
+import { useEffect, useState } from "react";
+import { DeviceInfo } from "../../common/DeviceManager";
+import { useDevices } from "../providers/DevicesProvider";
+import Button from "./shared/Button";
+import { useModal } from "../providers/ModalProvider";
+import { formatDisplayName, MAX_DEVICE_NAME_LENGTH } from "../views/CreateDeviceView";
+import Label from "../components/shared/Label";
+import { useProject } from "../providers/ProjectProvider";
+
+function DeviceRenameDialog({
+  deviceInfo,
+  onClose,
+}: {
+  deviceInfo: DeviceInfo;
+  onClose: () => void;
+}) {
+  const [displayName, setDisplayName] = useState<string>(deviceInfo.displayName);
+  const [isDisplayNameValid, setIsDisplayNameValid] = useState(true);
+  const { deviceManager } = useDevices();
+  const { project } = useProject();
+
+  const { showHeader } = useModal();
+  useEffect(() => {
+    showHeader(false);
+    return () => {
+      showHeader(true);
+    };
+  });
+
+  const handleDisplayNameChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const text = formatDisplayName(event.target.value);
+    setDisplayName(text);
+    setIsDisplayNameValid(0 < text.length && text.length <= MAX_DEVICE_NAME_LENGTH);
+  };
+
+  return (
+    <div>
+      <div className="device-rename-title">Rename device</div>
+      <div className="device-rename-wrapper">
+        <Label>
+          <span>New Name</span>
+        </Label>
+        <input
+          value={displayName}
+          className="device-name-input"
+          style={isDisplayNameValid ? {} : { border: "1px solid var(--red-light-100)" }}
+          type="string"
+          defaultValue={deviceInfo.displayName}
+          onChange={handleDisplayNameChange}
+        />
+        {!isDisplayNameValid && (
+          <div className="submit-rejection-message">
+            Make sure that the custom name is between 1 and {MAX_DEVICE_NAME_LENGTH} characters
+            long.
+          </div>
+        )}
+      </div>
+
+      <div className="device-rename-button-group">
+        <Button type="secondary" className="device-rename-button" onClick={onClose}>
+          Cancel
+        </Button>
+        <Button
+          className="device-rename-button"
+          type="ternary"
+          disabled={!isDisplayNameValid}
+          onClick={async () => {
+            const newDisplayName = displayName.trim();
+            try {
+              await deviceManager.renameDevice(deviceInfo, newDisplayName);
+            } finally {
+              deviceInfo.displayName = newDisplayName;
+              project.updateSelectedDevice(deviceInfo);
```

Comment:
I understand that you can update any device from the manage devices list. Why are we updating the selected one here then?

Reply:
Feels like this isn't the right place to update the selected device. First of all, it still seems to assume that we can only edit the selected device which is weird. We're also introducing a new method to project interface with unclear semantics (it actually changes the selected device info object but does not trigger the device change which can be very confusing).

I think the whole renaming of the current device should be entirely handled by the extension code and there should be no additional code for it in the UI. When someone renames the selected device, we can trigger `updateProjectState` from within the project and this should propagate down to the UI normally.

Reply:
I agree that the webview shouldn't trigger updateProjectState, the naming of updateSelectedDevice is unclear, and its purpose isn't intuitive. Instead updateSelectedDevice method, we should use the new method renameDevice also from project.ts, which will call renameDevice from DeviceManager.ts to save the name change and update the project state only if it’s the currently selected device that is being renamed.